lib/paloaltofw.py

```diff
@@ -0,0 +1,579 @@
+                                     "</member>")
+      address_group_entries.append(self.INDENT * 7 + "</static>")
+      address_group_entries.append(self.INDENT * 7 + "</entry>")
+    address_group_entries.append(self.INDENT * 5 + "</address-group>")
+
+    # SERVICES
+    service = []
+    service.append(self.INDENT * 5 + "<!-- Services -->")
+
+    service.append(self.INDENT * 5 + "<service>")
+    for k, v in Service.service_map.items():
+      service.append(self.INDENT * 6 + '<entry name="' + v["name"] + '">')
+      service.append(self.INDENT * 7 + "<protocol>")
+      service.append(self.INDENT * 8 + "<" + k[1] + ">")
+      tup = str(k[0])[1:-1]
+      if tup[-1] == ",":
+        tup = tup[:-1]
+      service.append(self.INDENT * 9 + "<port>" + tup.replace("'", "") +
+                     "</port>")
+      service.append(self.INDENT * 8 + "</" + k[1] + ">")
+      service.append(self.INDENT * 7 + "</protocol>")
+      service.append(self.INDENT * 6 + "</entry>")
+    service.append(self.INDENT * 5 + "</service>")
+
+    # RULES
+    rules = []
+    rules.append(self.INDENT * 5 + "<!-- Rules -->")
+
+    rules.append(self.INDENT * 5 + "<rulebase>")
+    rules.append(self.INDENT * 6 + "<security>")
+    rules.append(self.INDENT * 7 + "<rules>")
+
+    for name, options in Rule.rules.items():
+      rules.append(self.INDENT * 8 + '<entry name="' + name + '">')
+
+      rules.append(self.INDENT * 9 + "<to>")
+      for tz in options["to_zone"]:
+        rules.append(self.INDENT * 10 + "<member>" + tz + "</member>")
+      rules.append(self.INDENT * 9 + "</to>")
+
+      rules.append(self.INDENT * 9 + "<from>")
+      for fz in options["from_zone"]:
+        rules.append(self.INDENT * 10 + "<member>" + fz + "</member>")
+      rules.append(self.INDENT * 9 + "</from>")
+
+      rules.append(self.INDENT * 9 + "<source>")
+      if not options["source"]:
+        rules.append(self.INDENT * 10 + "<member>any</member>")
+      else:
+        for s in options["source"]:
+          rules.append(self.INDENT * 10 + "<member>" + s + "</member>")
+      rules.append(self.INDENT * 9 + "</source>")
+
+      rules.append(self.INDENT * 9 + "<destination>")
+      if not options["destination"]:
+        rules.append(self.INDENT * 10 + "<member>any</member>")
+      else:
+        for d in options["destination"]:
+          rules.append(self.INDENT * 10 + "<member>" + d + "</member>")
+      rules.append(self.INDENT * 9 + "</destination>")
+
+      rules.append(self.INDENT * 9 + "<service>")
+      if not options["service"]:
+        rules.append(self.INDENT * 10 + "<member>any</member>")
+      else:
+        for s in options["service"]:
+          rules.append(self.INDENT * 10 + "<member>" + s + "</member>")
+      rules.append(self.INDENT * 9 + "</service>")
+
+      rules.append(self.INDENT * 9 + "<action>" + Term.ACTIONS.get(
+          str(options["action"])) + "</action>")
+
+      if fz == tz == "any":
```

Comment:
This seems weird. There are loops above over the from zone and to zone which is a list. From my bit of testing I don't see that this would ever have more than one from-zone or to-zone per rule in rules. So why are these added as a list? The way the logic is setup fz could theoretically be set to an empty list if someone fed in a modified policy object that had the options set to ['from-zone', '', 'to-zone', ''] this would cause a problem.